src/linear_algebra/finsupp.lean

```diff
@@ -141,247 +101,252 @@ begin
-iff.rfl
+variables {β}
 
-section
-set_option class.instance_max_depth 37
-def restrict_dom_equiv_finsupp (s : set α) [decidable_pred (λ (x : α), x ∈ s)] :
-  (restrict_dom β γ s) ≃ₗ[γ] (s →₀ β) :=
-(restrict_support_equiv s).to_linear_equiv
-begin
-  show is_linear_map γ ((lsubtype_domain s : (α →₀ β) →ₗ[γ] (s →₀ β)).comp
-      (submodule.subtype (restrict_dom β γ s))),
-  exact linear_map.is_linear _
-end
-end
+lemma mem_supported {s : set α} (p : α →₀ β) : p ∈ (supported β γ s) ↔ ↑p.support ⊆ s :=
+iff.rfl
 
-end
+lemma mem_supported' {s : set α} (p : α →₀ β) :
+  p ∈ supported β γ s ↔ ∀ x ∉ s, p x = 0 :=
+by simp [mem_supported, set.subset_def, not_imp_comm]
 
-end module
+lemma single_mem_supported {s : set α} {a : α} (b : β) (h : a ∈ s) :
+  single a b ∈ supported β γ s :=
+set.subset.trans support_single_subset (set.singleton_subset_iff.2 h)
 
-section vector_space
-variables {α : Type*} {β : Type*} {γ : Type*}
-variables [decidable_eq α] [decidable_eq β] [discrete_field γ] [add_comm_group β] [vector_space γ β]
+lemma supported_eq_span_single [has_one β] (s : set α) :
+  supported γ γ s = span γ ((λ i, single i 1) '' s) :=
+begin
+  refine (span_eq_of_le _ _ (le_def'.2 $ λ l hl, _)).symm,
+  { rintro _ ⟨_, hp, rfl ⟩ , exact single_mem_supported γ 1 hp },
+  { rw ← l.sum_single,
+    refine sum_mem _ (λ i il, _),
+    convert @smul_mem γ (α →₀ γ) _ _ _ _ (single i 1) (l i) _,
+    { simp },
+    apply subset_span,
+    apply set.mem_image_of_mem _ (hl il) }
+end
 
-open linear_map submodule
+variables (β γ)
 
-instance : vector_space γ (α →₀ β) :=
-{ to_module := finsupp.module }
+def restrict_dom (s : set α) : α →₀ β →ₗ supported β γ s :=
+linear_map.cod_restrict _
+  { to_fun := filter (∈ s),
+    add := λ l₁ l₂, filter_add,
+    smul := λ a l, filter_smul }
+  (λ l, (mem_supported' _ _).2 $ λ x, filter_apply_neg (∈ s) l)
 
-lemma is_basis_single {f : α → set β} (hf : ∀a, is_basis γ (f a)) :
-  is_basis γ (⋃a, single a '' f a) :=
-⟨linear_independent_single $ assume a, (hf a).1,
-  by simp only [span_Union, span_single_image, (hf _).2, map_top, supr_lsingle_range]⟩
+variables {β γ}
 
-end vector_space
+set_option class.instance_max_depth 50
+@[simp] theorem restrict_dom_apply (s : set α) (l : α →₀ β) :
+  ↑((restrict_dom β γ s : α →₀ β →ₗ supported β γ s) l) = finsupp.filter (∈ s) l := rfl
+set_option class.instance_max_depth 32
 
-section dim
-universes u v
-variables {α : Type u} {β : Type u} {γ : Type v}
-variables [decidable_eq α] [decidable_eq β] [discrete_field γ] [add_comm_group β] [vector_space γ β]
+theorem restrict_dom_comp_subtype (s : set α) :
+  (restrict_dom β γ s).comp (submodule.subtype _) = linear_map.id :=
+by ext l; apply subtype.coe_ext.2; simp; ext a;
+   by_cases a ∈ s; simp [h]; exact ((mem_supported' γ l.1).1 l.2 a h).symm
 
-lemma dim_eq : vector_space.dim γ (α →₀ β) = cardinal.mk α * vector_space.dim γ β :=
+theorem range_restrict_dom (s : set α) : (restrict_dom β γ s).range = ⊤ :=
```

Comment:
(general comment) I notice that you have a `finsupp_vector_space` file now. I think most of these theorems should go there instead of here, which is supposed to be a more basic file. How about renaming it to `finsupp_module` and moving all the linear algebra stuff there, i.e. everything that depends on `linear_algebra.basic`, and remove the dependence of this file on `linear_algebra.basic`?

Reply:
I created `finsupp_vector_space` because it depends on `dimension` and `basis`, which in turn depends on `finsupp`. So I have to keep the two separate.